File: C++/check-if-strings-can-be-made-equal-with-operations-i.cpp
// Time:  O(1)
// Space: O(1)

// freq table
class Solution {
public:
    bool canBeEqual(string s1, string s2) {
        vector<vector<int>> cnt1(2, vector<int>(26)), cnt2(2, vector<int>(26));
        for (int i = 0; i < 2; ++i) {
            for (int j = i; j < size(s1); j += 2) {
                ++cnt1[i][s1[j] - 'a'];
            }
            for (int j = i; j < size(s1); j += 2) {
                ++cnt2[i][s2[j] - 'a'];
            }
        }
        return cnt1 == cnt2;
    }
};

// Time:  O(1)
// Space: O(1)
// brute force
class Solution2 {
public:
    bool canBeEqual(string s1, string s2) {
        return ((s1[0] == s2[0] && s1[2] == s2[2]) || (s1[0] == s2[2] && s1[2] == s2[0])) &&
               ((s1[1] == s2[1] && s1[3] == s2[3]) || (s1[1] == s2[3] && s1[3] == s2[1]));
    }
};

File: Python/check-if-strings-can-be-made-equal-with-operations-i.py
# Time:  O(1)
# Space: O(1)

import collections


# freq table
class Solution(object):
    def canBeEqual(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        return all(collections.Counter(s1[j] for j in xrange(i, len(s1), 2)) == collections.Counter(s2[j] for j in xrange(i, len(s2), 2)) for i in xrange(2))


# Time:  O(1)
# Space: O(1)
# brute force
class Solution2(object):
    def canBeEqual(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        return (((s1[0] == s2[0] and s1[2] == s2[2]) or (s1[0] == s2[2] and s1[2] == s2[0])) and
                ((s1[1] == s2[1] and s1[3] == s2[3]) or (s1[1] == s2[3] and s1[3] == s2[1])))
